core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java

```diff
@@ -0,0 +1,198 @@
+        ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> toReport = MissingClassTelemetry.events.getEventsAndClean();
+        if (LOGGER.isLoggable(Level.FINE)) LOGGER.fine("Cleaned events for missed classes on telemetry for Java 11");
+
+        return formatEvents(toReport);
+    }
+
+    /**
+     * Format the events gathered in a map used to create the json object to send via telemetry. The events are named by
+     * the class not found. But a class could be repeated if it was thrown from several places. The interesting
+     * pieces of information we want to gather are the places where the {@link ClassNotFoundException} or the
+     * {@link NoClassDefFoundError} errors happens, rather than the class itself.
+     * @param events events collected in this telemetry window.
+     * @return the events formatted in a map.
+     */
+    @Nonnull
+    private JSONArray formatEvents(@Nonnull ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> events) {
+        JSONArray jsonEvents = new JSONArray();
+
+        events.forEach((stackTrace, event) -> {
+            JSONObject eventObject = new JSONObject();
+            eventObject.put("className", event.getClassName());
+            eventObject.put("class", event.getClassName());
+            eventObject.put("time", event.getTime());
+            eventObject.put("occurrences", Long.toString(event.getOccurrences()));
+            eventObject.put("stacktrace", event.getStackTrace());
+
+            jsonEvents.add(eventObject);
+        });
+
+        return jsonEvents;
+    }
+
+    /**
+     * The current time in the same way as other telemetry implementations.
+     * @return the UTC time formatted with the pattern yyyy-MM-dd'T'HH:mm'Z'
+     */
+    @Nonnull
+    static String clientDateString() {
+        TimeZone tz = TimeZone.getTimeZone("UTC");
+        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'");
+        df.setTimeZone(tz); // strip timezone
+        return df.format(new Date());
+    }
+
+    /**
+     * Store the exception if it's from a split package of Java.
+     * @param name the name of the class
+     * @param e the exception thrown
+     */
+    public static void reportException(@Nonnull String name, @Nonnull Throwable e) {
+        if (isFromMovedPackage(name)) {
+            events.put(e);
+
+            if (LOGGER.isLoggable(Level.FINE)) LOGGER.log(Level.FINE, "Added a missed class for Java 11 telemetry. Class: " + name, e);
+        }
+    }
+
+    private static boolean isFromMovedPackage(@Nonnull String clazz) {
+        for (String movedPackage : MOVED_PACKAGES) {
+            if (clazz.startsWith(movedPackage)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * List of exception classes we are going to send via telemetry.
+     * @return the list of exception classes
+     */
+    public static Class[] getCollectibleThrowables() {
```

Comment:
Are we sure this will always be the same classloader? IOW, that the match above will work?

Because `Class.equals()` will be true only if the class is the same AND they have been loaded by the same classloader. Which I suspect might not be always true, and hence we'd miss some exceptions?

Reply:
I got your point, but It's catching exceptions raised by the core itself and from a tester plugin, so it seems like nothing is escaping. But after your comment, I've found out that I was only catching exceptions wrapped in a `ServletException` directly because I was getting the cause directly. I've changed that to look into the exception itself, the causes recursively and even the suppressed exceptions.